File: src/link/log/log.cpp
// File: log.cpp
// Project: log
// Created Date: 27/01/2023
// Author: Shun Suzuki
// -----
// Last Modified: 17/04/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include "autd3/link/log.hpp"

#include "../../spdlog.hpp"
#include "autd3/core/link.hpp"
#include "autd3/extra/cpu_emulator.hpp"

namespace autd3::link {

class LogImpl final : public core::Link {
 public:
  explicit LogImpl(core::LinkPtr link, std::shared_ptr<spdlog::logger> logger) : Link(), _link(std::move(link)), _logger(std::move(logger)) {}
  ~LogImpl() override = default;
  LogImpl(const LogImpl& v) noexcept = delete;
  LogImpl& operator=(const LogImpl& obj) = delete;
  LogImpl(LogImpl&& obj) = delete;
  LogImpl& operator=(LogImpl&& obj) = delete;

  bool open(const core::Geometry& geometry) override {
    _synchronized = false;
    _logger->debug("Open Log link");
    if (is_open()) {
      _logger->warn("Link is already opened");
      return true;
    }
    if (!_link->open(geometry)) {
      _logger->error("Failed to open link");
      return false;
    }

    return true;
  }

  bool close() override {
    _synchronized = false;

    _logger->debug("Close Log link");

    if (!is_open()) {
      _logger->warn("Link is not opened");
    }
    if (!_link->close()) {
      _logger->error("Failed to close link");
      return false;
    }

    return true;
  }

  bool send(const driver::TxDatagram& tx) override {
    _logger->debug("Send data");
    if (!is_open()) {
      _logger->warn("Link is closed");
      return false;
    }

    switch (tx.header().msg_id) {
      case driver::MSG_CLEAR:
      case driver::MSG_RD_CPU_VERSION_MAJOR:
      case driver::MSG_RD_FPGA_VERSION_MAJOR:
      case driver::MSG_RD_CPU_VERSION_MINOR:
      case driver::MSG_RD_FPGA_VERSION_MINOR:
      case driver::MSG_RD_FPGA_FUNCTION:
        break;
      default:
        if (!tx.header().cpu_flag.contains(driver::CPUControlFlags::ConfigEnN) && tx.header().cpu_flag.contains(driver::CPUControlFlags::ConfigSync))
          _synchronized = true;
        if (!_synchronized) _logger->warn("Devices are not synchronized!");
        break;
    }

    if (!_link->send(tx)) {
      _logger->error("Failed to send data");
      return false;
    }

    return true;
  }

  bool receive(driver::RxDatagram& rx) override {
    _logger->debug("Receive data");
    if (!is_open()) {
      _logger->warn("Link is closed");
      return false;
    }

    if (!_link->receive(rx)) {
      _logger->error("Failed to receive data");
      return false;
    }

    return true;
  }

  bool send_receive(const driver::TxDatagram& tx, driver::RxDatagram& rx, const std::optional<core::Duration> timeout) override {
    if (!send(tx)) return false;
    const auto timeout_ = timeout.value_or(_timeout);
    if (timeout_ == core::Duration::zero()) return receive(rx);
    if (!wait_msg_processed(tx.header().msg_id, rx, timeout_)) {
      _logger->error("Failed to confirm that the data was processed");
      return false;
    }
    return true;
  }

  bool is_open() override { return _link->is_open(); }

 private:
  bool _synchronized{false};
  core::LinkPtr _link;
  std::vector<extra::CPU> _cpus;
  std::shared_ptr<spdlog::logger> _logger;
};

core::LinkPtr Log::build() {
  const auto name = "AUTD3 Log";
  spdlog::sink_ptr sink =
      _out == nullptr || _flush == nullptr ? get_default_sink() : std::make_shared<CustomSink<std::mutex>>(std::move(_out), std::move(_flush));
  auto logger = std::make_shared<spdlog::logger>(name, std::move(sink));
  logger->set_level(static_cast<spdlog::level::level_enum>(_level));

  core::LinkPtr link = std::make_unique<LogImpl>(std::move(_link), std::move(logger));
  return link;
}

}  // namespace autd3::link

File: src/autd3/src/link/log.rs
/*
 * File: log.rs
 * Project: link
 * Created Date: 10/05/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 04/10/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */

use std::{
    marker::PhantomData,
    ops::{Deref, DerefMut},
    time::Duration,
};

use autd3_driver::{
    cpu::{RxMessage, TxDatagram},
    error::AUTDInternalError,
    geometry::{Device, Transducer},
    link::Link,
};
use spdlog::prelude::*;

pub use spdlog::Level;

/// Link to logging
pub struct Log<T: Transducer, L: Link<T>> {
    link: L,
    logger: Logger,
    _trans: PhantomData<T>,
}

pub trait IntoLog<T: Transducer, L: Link<T>> {
    /// enable logger
    fn with_log(self) -> Log<T, L>;
}

impl<T: Transducer, L: Link<T>> IntoLog<T, L> for L {
    fn with_log(self) -> Log<T, L> {
        Log::new(self)
    }
}

impl<T: Transducer, L: Link<T>> Deref for Log<T, L> {
    type Target = L;

    fn deref(&self) -> &Self::Target {
        &self.link
    }
}

impl<T: Transducer, L: Link<T>> DerefMut for Log<T, L> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.link
    }
}

impl<T: Transducer, L: Link<T>> Log<T, L> {
    fn new(link: L) -> Self {
        let logger = autd3_driver::logger::get_logger();
        logger.set_level_filter(spdlog::LevelFilter::MoreSevereEqual(spdlog::Level::Info));
        Self {
            link,
            logger,
            _trans: PhantomData,
        }
    }

    /// set log level
    pub fn with_log_level(self, level: LevelFilter) -> Self {
        self.logger.set_level_filter(level);
        self
    }

    /// set logger
    ///
    /// By default, the logger will display the log on the console.
    pub fn with_logger(self, logger: Logger) -> Self {
        Self { logger, ..self }
    }
}

impl<T: Transducer, L: Link<T>> Link<T> for Log<T, L> {
    fn open(&mut self, devices: &[Device<T>]) -> Result<(), AUTDInternalError> {
        trace!(logger: self.logger, "Open Log link");

        if self.is_open() {
            warn!(logger: self.logger, "Log link is already opened.");
            return Ok(());
        }

        let res = self.link.open(devices);
        if res.is_err() {
            error!(logger: self.logger, "Failed to open link");
            return res;
        }

        Ok(())
    }

    fn close(&mut self) -> Result<(), AUTDInternalError> {
        trace!(logger: self.logger, "Close Log link");

        if !self.is_open() {
            warn!(logger: self.logger, "Log link is already closed.");
            return Ok(());
        }

        let res = self.link.close();
        if res.is_err() {
            error!(logger: self.logger, "Failed to close link");
            return res;
        }

        Ok(())
    }

    fn send(&mut self, tx: &TxDatagram) -> Result<bool, AUTDInternalError> {
        debug!(logger: self.logger, "Send data");

        if !self.is_open() {
            warn!(logger: self.logger, "Link is not opened");
            return Ok(false);
        }

        if !self.link.send(tx)? {
            error!(logger: self.logger, "Failed to send data");
            return Ok(false);
        }
        Ok(true)
    }

    fn receive(&mut self, rx: &mut [RxMessage]) -> Result<bool, AUTDInternalError> {
        debug!(logger: self.logger, "Receive data");

        if !self.is_open() {
            warn!(logger: self.logger, "Link is not opened");
            return Ok(false);
        }

        if !self.link.receive(rx)? {
            error!(logger: self.logger, "Failed to receive data");
            return Ok(false);
        }
        Ok(true)
    }

    fn send_receive(
        &mut self,
        tx: &TxDatagram,
        rx: &mut [RxMessage],
        timeout: Duration,
    ) -> Result<bool, AUTDInternalError> {
        if !self.send(tx)? {
            return Ok(false);
        }
        if timeout.is_zero() {
            return self.receive(rx);
        }
        if !self.wait_msg_processed(tx, rx, timeout)? {
            error!(logger: self.logger, "Failed to confirm that the data was processed");
            return Ok(false);
        }
        Ok(true)
    }

    fn is_open(&self) -> bool {
        self.link.is_open()
    }

    fn timeout(&self) -> Duration {
        Duration::ZERO
    }
}
